File: Javascript/2.Settembre/37.SentenceCapitalize-16Settembre.js
//todo Dato un paragrafo, restituisci un nuovo paragrafo in cui la prima lettera di ogni frase è maiuscola.


//! Tutti gli altri caratteri devono essere mantenuti.
//! Le frasi possono terminare con un punto (.), uno o più punti interrogativi (?) o uno o più punti esclamativi (!).

//? Tempo di esecuzione:
//* Non completato 

//?Codice scritto senza utilizzo di ChatGpt

function capitalize(paragraph) {
  
  let arrParagr = paragraph.split("")

  let word = [];

  for(let i = 0; i < arrParagr.length; i++) {
    if(i === 0) word.push(arrParagr[i].toUpperCase())
    if(arrParagr[i] === "." || 
       arrParagr[i] === "?" || 
       arrParagr[i] === "!" ) word.push(arrParagr[i+2].toUpperCase())
  }
  return word;
}

console.log(capitalize("hello world. how are you?"))

//?Codice scritto da ChatGpt

function capitalize(paragraph) {
  let result = "";
  let capitalizeNext = true;

  for (let i = 0; i < paragraph.length; i++) {
    const char = paragraph[i];

    if (capitalizeNext && /[a-zA-Z]/.test(char)) {
      result += char.toUpperCase();
      capitalizeNext = false;
    } else {
      result += char;
    }

    if (char === "." || char === "?" || char === "!") {
      capitalizeNext = true;
    }
  }

  return result;
}

console.log(capitalize("hello world. how are you?"));

//* ✅ Pro

//* Mantiene tutti i caratteri originali (spazi, apostrofi, punteggiatura)

//* Gestisce più segni consecutivi (..., !!!, ???)

//* Logica chiara: capitalizza solo quando serve

//* Passa tutti i test

//* Codice facile da leggere e mantenere

//! ❌ Contro

//! Usa una regex (/[a-zA-Z]/) → leggermente più “avanzata”

//! Scorre l’intera stringa carattere per carattere (ok per testi normali)

//! Un po’ più verboso rispetto a soluzioni “furbe” con split

//todo Considerazioni: 

//* Mi sono appena svegliato e non sono al 100% delle mie capacità logiche.

//! N.B.

File: Python/2.Settembre/37.Sentence_capitalize-16Settembre.py
# =========================================================
# FILE TEMPLATE - ANALISI ESERCIZI PYTHON
# =========================================================

# //Prompt ChatGpt: questo è il modello. Invierò consegna esercizio e codice, e vorrei feedback strutturato

# =========================================================
# //todo Consegna dell'esercizio
# =========================================================

# Dato un paragrafo, restituisci un nuovo paragrafo in cui la prima
# lettera di ogni frase è maiuscola.
#
# Tutti gli altri caratteri devono essere mantenuti.
#
# Le frasi possono terminare con:
# - un punto (.)
# - uno o più punti interrogativi (?)
# - uno o più punti esclamativi (!)

# =========================================================
# //* Codice scritto senza utilizzo di ChatGpt
# =========================================================

import re

def capitalize(testo):
    if not testo:
        return testo

    lettere = r"[a-zA-ZàèìòùÀÈÌÒÙáéíóúÁÉÍÓÚ]"

    testo = re.sub(f"({lettere})", lambda m: m.group(1).upper(), testo, count=1)

    pattern = f"([\\.!\?]+)(\\s*)({lettere})"

    risultato = re.sub(
        pattern,
        lambda m: m.group(1) + m.group(2) + m.group(3).upper(),
        testo
    )

    return risultato

stringa_test = "crazy!!!strange???unconventional...sentences."
print(capitalize(stringa_test))

# =========================================================
# //* Codice scritto da ChatGpt (versione migliorata)
# =========================================================

import re

def capitalize_sentences(text):
    if not text:
        return text

    result = []
    capitalize_next = True

    for char in text:
        if capitalize_next and char.isalpha():
            result.append(char.upper())
            capitalize_next = False
        else:
            result.append(char)

        if char in ".!?":
            capitalize_next = True

    return "".join(result)

# =========================================================
# //* Pro e contro del mio codice
# =========================================================

# //* Pro
# - Utilizza le regex in modo corretto.
# - Gestisce punti, punti interrogativi e punti esclamativi multipli.
# - Mantiene invariati gli altri caratteri.
# - Supporta lettere accentate.
# - Soluzione compatta e leggibile per chi conosce le regex.

# //! Contro
# - Uso delle regex più complesso del necessario.
# - La variabile "lettere" richiede manutenzione se si vogliono supportare
#   altri alfabeti o caratteri Unicode.
# - Meno immediato da comprendere rispetto a una soluzione iterativa.
# - Esegue due sostituzioni regex complete sul testo.

# =========================================================
# //* Consigli
# =========================================================

# - La tua soluzione è valida e rispetta la consegna.
# - Hai pensato correttamente al caso delle frasi senza spazi dopo
#   la punteggiatura ("ciao!!!come va?").
# - Ottimo l'uso di (\s*) per gestire sia la presenza che l'assenza
#   di spazi tra le frasi.
# - Per un colloquio tecnico probabilmente la tua soluzione sarebbe
#   considerata più avanzata della media.
# - Se l'obiettivo è massima leggibilità, una soluzione iterativa
#   senza regex potrebbe risultare più semplice da mantenere.
# - Complessivamente darei al codice un voto di 8.5/10. Funziona bene,
#   è robusto e mostra una buona padronanza delle espressioni regolari.